### DISCUSSION TITLE
How to build in some .exe file?

### DISCUSSION THREAD

**Discussion by Samwellwang** [flet-dev/flet#3787] [Q&A]
### Duplicate Check

- [X] I have searched the [opened issues](https://github.com/flet-dev/flet/issues) and there are no duplicates

### Describe the requested feature

eg, I want to build ffmpeg to my windows app . What should I do.

**Comment by baseplate-admin** [comment]
I am in this exact same phase.

Use a helper script to copy the ffmpeg to flet directory

**Comment by Samwellwang** [comment]
> I am in this exact same phase.
> 
> Use a helper script to copy the ffmpeg to flet directory

Hi ,Could you please describe it in more detail? I put ffmpeg.exe in flet directory already. But when I bulit with "flet build windows" and then run my app . It shows that can’t find ffmpeg.exe in "C://User/xxx/AppData/Local/XX" which is flutter build temp dir.

**Comment by Samwellwang** [comment]
> > I am in this exact same phase.
> > Use a helper script to copy the ffmpeg to flet directory
> 
> Hi ,Could you please describe it in more detail? I put ffmpeg.exe in flet directory already. But when I bulit with "flet build windows" and then run my app . It shows that can’t find ffmpeg.exe in "C://User/xxx/AppData/Local/XX" which is flutter build temp dir.

I found the old build command “flet pack xxx.py  ” is useful for me ==!

**Comment by baseplate-admin** [comment]
There is another way i think, i haven't tested it. But you can put your exe in `assets/` directory and call it a day.

**Comment by Samwellwang** [comment]
> But you can put your exe in `assets/`

Thanks for you reply, I put the fmmpeg.exe in assets/  dir ,but I always get the temp path of building exe  path (not real path with exe) in my app ,which is built use "flet build windows"

**Comment by baseplate-admin** [comment]
so bascially you have to CD into the assets directory.

Something like this might be doable by doing

```python
from pathlib import Path
import os 

BASE_DIR = Path(__file__).resolve().parent


ASSETS_DIR = os.path.join(BASE_DIR, 'assets')

FFMPEG_BINARY = os.path.join(ASSETS_DIR, 'ffmpeg.exe')
```

**Comment by Samwellwang** [comment]
> so bascially you have to CD into the assets directory.
> 
> Something like this might be doable by doing
> 
> ```python
> from pathlib import Path
> import os 
> 
> BASE_DIR = Path(__file__).resolve().parent
> 
> 
> ASSETS_DIR = os.path.join(BASE_DIR, 'assets')
> 
> FFMPEG_BINARY = os.path.join(ASSETS_DIR, 'ffmpeg.exe')
> ```

Unfortunately，，It's didn't work for me 
![image](https://github.com/user-attachments/assets/b37dee5c-067e-4452-a108-6cd35d4916a7)

**Comment by baseplate-admin** [comment]
> Unfortunately，，It's didn't work for me

However it did work with my local build. could you tree your build directory and your project directory?

**Comment by Samwellwang** [comment]
> could you tree your build directory and your project directory?
This is tree of  my project directory.
/assets/
     icon.png
     ffmpeg/
         bin/
             ffmpeg.exe
             ffprobe.exe
             ffplay.exe
main.py
requirements.txt
pyproject.toml

**Comment by baseplate-admin** [comment]
Okay, i will try to setup a demo project, where i will try to debug the issue.

**Comment by DrRoot-github** [comment]
Here’s a workaround I found and would like to share:
Since the Python executable runs from "serious-python" in the AppData directory after the build, you can use psutil to locate the path of the exe file. This allows you to handle paths in both development and production environments.

```python
class DirectoryUtil:
    exe_name = "yourexefile.exe"
    base_dir = None

    @classmethod
    def init_base_directory(cls):
        # when running serious-python, sys.argv is empty([])
        if not sys.argv[0]:
            for pid in psutil.pids():
                proc = psutil.Process(pid)
                if proc.name() == cls.exe_name:
                    cls.base_dir = os.path.dirname(proc.exe())
        else:
            cls.base_dir = os.getcwd()
```
